lib/Sema/CSDiag.cpp

```diff
@@ -980,6 +981,69 @@ static bool argumentMismatchIsNearMiss(Type argType, Type paramType) {
   return false;
 }
 
+/// Given a parameter type that may contain generic type params and an actual
+/// argument type, decide whether the param and actual arg have the same shape
+/// and equal fixed type portions, and return by reference each archetype and
+/// the matching portion of the actual arg type where that archetype appears.
+static bool findGenericSubstitutions(Type paramType, Type actualArgType,
+                                     SmallVector<ArchetypeType *, 4> &archetypes,
+                                     SmallVector<Type, 4> &substitutions) {
+  if (auto genericParam = paramType->getAs<GenericTypeParamType>())
+    paramType = genericParam->getDecl()->getArchetype();
+  if (auto archetype = paramType->getAs<ArchetypeType>()) {
+    for (unsigned i = 0, c = archetypes.size(); i < c; i++) {
+      if (archetypes[i]->isEqual(archetype))
+        return substitutions[i]->isEqual(actualArgType);
+    }
+    archetypes.push_back(archetype);
+    substitutions.push_back(actualArgType);
+    return true;
+  }
+  
+  if (paramType->isEqual(actualArgType))
+    return true;
+  if (paramType->getKind() != actualArgType->getKind())
+    return false;
+  
+  // A little walker that just collects direct child types. Can't easily have a
+  // walker walk recursively because we need to coordinate two of them.
+  struct ChildWalker : public TypeWalker {
+  public:
+    SmallVector<Type, 4> children;
+    bool top = true;
+    
+    virtual Action walkToTypePre(Type ty) {
+      if (top) {
+        top = false;
+        return Action::Continue;
+      }
+      children.push_back(ty);
+      return Action::SkipChildren;
+    }
+  } paramWalker, actualArgWalker;
+  
+  paramType.walk(paramWalker);
+  actualArgType.walk(actualArgWalker);
+  
+  unsigned childCount = paramWalker.children.size();
+  if (childCount == 0 || actualArgWalker.children.size() != childCount)
+    return false;
+  for(unsigned i = 0; i < childCount; i++) {
+    if (!findGenericSubstitutions(paramWalker.children[i], actualArgWalker.children[i],
+                                  archetypes, substitutions))
+      return false;
+  }
+  
+  // Unfortunate special case, FunctionType also needs to compare ExtInfo. This is the only
```

Comment:
Without this check, test/attr/attr_autoclosure.swift:53 regresses and emits the old worse diagnoses from before https://github.com/apple/swift/pull/1042:
```
func overloadedEach<O: P1>(source: O, _ closure: () -> ()) {
}

func overloadedEach<P: P2>(source: P, _ closure: () -> ()) {
}

struct S : P2 {
  typealias Element = Int
  func each(@autoclosure closure: () -> ()) {
    overloadedEach(self, closure) // expected-error {{invalid conversion from non-escaping function of type '@autoclosure () -> ()' to potentially escaping function type '() -> ()'}}
  }
}
```